### ISSUE TITLE
Azure credentials provider lowercase parameters

### ISSUE THREAD

**Issue by Beesblaas** [aws/amazon-redshift-python-driver#49]
## Driver version
2.0.883

## Redshift version
PostgreSQL 8.0.2 on i686-pc-linux-gnu, compiled by GCC gcc (GCC) 3.4.2 20041017 (Red Hat 3.4.2-6.fc3), Redshift 1.0.28917

## Client Operating System
Ubuntu 20.04.2 LTS

## Python version
Python 3.8.10

## Table schema
N/A
## Problem description
Connection to Redshift using AzureCredentialsProvider

1. Expected behaviour: Connection successful.
2. Actual behaviour: Fails with Invalid parameter value.
3. Error message/stack trace: ClientError: An error occurred (InvalidParameterValue) when calling the GetClusterCredentials operation: DbUser parameter must be lowercase, begin with a letter, contain only alphanumeric characters, underscore ('_'), plus sign ('+'), dot ('.'), at ('@'), or hyphen ('-'), and be less than 128 characters.
4. Any other details that can be helpful: Looks like connection was failing because the DbUser parameter needed to be lowercase, but it was already lowercase. Then found the /usr/local/lib/python3.8/dist-packages/redshift_connector/iam_helper.py file and updated the below to test, by forcing the parameter "DbUser " to be lowercase. 

```python
cred = typing.cast(
                    typing.Dict[str, typing.Union[str, datetime.datetime]],
                    client.get_cluster_credentials(
                        DbUser=str(info.db_user).lower(),
                        DbName=info.db_name,
                        DbGroups=info.db_groups,
                        ClusterIdentifier=info.cluster_identifier,
                        AutoCreate=info.auto_create,
                    ),
                )
```
This fixed the connection. Probably not the correct place to fix it but somewhere it is changing the case of the parameter I believe. Getting the same errors when using DBName, and DBGroups parameters but we did not require them.

## Reproduction code
```python
import redshift_connector
import pandas as pd

with open('untitled.txt') as f:
    lines = f.readlines()
    
cnxn = redshift_connector.connect(
    iam=True,
    ssl=True,
    host='',
    port=5439,
    database='dev',
    db_user='username in lowercase',
    cluster_identifier='',
    region='',
    credentials_provider='AzureCredentialsProvider',
    idp_tenant='',
    client_id='',
    client_secret='',
    user='',
    password=lines[0],
    preferred_role='arn:aws:iam::account_number:role/role_name',
    auto_create=True
)

QueryString = 'SELECT * FROM INFORMATION_SCHEMA.TABLES'
df1 = pd.read_sql(QueryString, cnxn)
table_names = list(df1[df1[b'table_type']=='BASE TABLE'][b'table_name'])
cnxn.close()
print(table_names)
```

**Comment by Brooke-white** [comment]
Hi @Beesblaas ,

Thank you for reaching out. The team will work on reproducing this issue and determining its cause.

**Comment by Brooke-white** [comment]
Hi @Beesblaas ,

In the code block you provided above,

```python
                    typing.Dict[str, typing.Union[str, datetime.datetime]],
                    client.get_cluster_credentials(
                        DbUser=str(info.db_user).lower(),
                        DbName=info.db_name,
                        DbGroups=info.db_groups,
                        ClusterIdentifier=info.cluster_identifier,
                        AutoCreate=info.auto_create,
                    ),
                )
```

If the value of `info.db_user` is not the same value you provided via the `db_user` connection parameter, that means it is being set to a different value while processing the SAML response from the given Azure enterprise application.

When taking a closer look at the code which overrides `RedshiftProperty` values based on the SAML response contents, I noticed some suspicious code where it appears sequential `if` statements are used rather than `if...elif`. This code appears [here](https://github.com/aws/amazon-redshift-python-driver/blob/v2.0.883/redshift_connector/iam_helper.py#L324..L336)

I anticipate the below correction should resolve this issue

```python
                if allow_db_user_override is True:
                    if saml_db_user is not None:
                        info.db_user = saml_db_user
                    elif db_user is not None:
                        info.db_user = db_user
                    elif profile_db_user is not None:
                        info.db_user = profile_db_user
                else:
                    if db_user is not None:
                        info.db_user = db_user
                    elif profile_db_user is not None:
                        info.db_user = profile_db_user
                    elif saml_db_user is not None:
                        info.db_user = saml_db_user
```

Please let me know if this change resolves your issue.

**Comment by Beesblaas** [comment]
Hi @Brooke-white 

Thanks for the quick response. I updated it to ```elif``` and then removed the ```DbUser=str(info.db_user).lower()``` part back to ```DbUser=info.db_user```. And the connection worked.

I then further just tested by including ```db_groups``` in my connection parameters as we might use it later, but that also got the below error. If you want to have a further look into that.

ClientError: An error occurred (InvalidParameterValue) when calling the GetClusterCredentials operation: Each group name in the DbGroups parameter list must be lowercase, begin with a letter, contain only alphanumeric characters, underscore ('_'), plus sign ('+'), dot ('.'), at ('@'), or hyphen ('-'), and be less than 128 characters.

But that change resolves it for `db_user`.

**Comment by Brooke-white** [comment]
Hi @Beesblaas ,

Glad to hear the modifications above resolved the `DbUser` issue. We will include these changes in our next release, scheduled to occur August 16th. 

Thank you for sharing the `db_groups` related error. I was able to reproduce this.

There is a connection parameter `force_lowercase` that when set to `True` should force all `db_groups` values to be lowercase. I was planning to recommend this setting...but it seems this functionality is only partially implemented.

So, I've included changes I expect to fix this issue below, for both the case where the `db_groups` parameter is included via the connection parameters and when it's provided in the SAML response from the application used for authentication. I performed a small test passing the `db_groups` using alternating upper and lower case and saw no issue:

`connect(...,db_groups=['eVeRyOnE', 'mYtEsTgRoUp'], force_lowercase=True)`

As before, these changes occur in the `iam_helper.py` file:

At [this](https://github.com/aws/amazon-redshift-python-driver/blob/v2.0.883/redshift_connector/iam_helper.py#L267) location I added the following:

```python
        if info.db_groups and info.force_lowercase:
            info.db_groups = [group.lower() for group in info.db_groups])
```

[this](https://github.com/aws/amazon-redshift-python-driver/blob/v2.0.883/redshift_connector/iam_helper.py#L338...L339) code block was modified to:

```python
                if (len(info.db_groups) == 0) and (len(db_groups) > 0):
                    if force_lowercase:
                        info.db_groups = [group.lower() for group in db_groups])
                    else:
                        info.db_groups = db_groups
```

Please let me know if this change resolves the issue at hand. Thank you for your willingness to dive into the codebase :)

**Comment by Beesblaas** [comment]
Hi @Brooke-white 

Sorry for only getting back now, and thanks for the quick responses. I am glad to assist where possible. I tested your change in regards to the  ```db_groups``` parameter and that worked as well. But  I realize I also have to use square brackets in the parameters for it to work, even if just passing one group as its a list. Missed that originally. 

````python
db_groups=['aws_aps_poc'],
````

Let me know if you require anymore testing, happy to assist.

**Comment by Brooke-white** [comment]
Hey @Beesblaas ,

Awesome glad to hear that worked out :). We will make the db_groups fix a part of our next release, scheduled for August 16th, as well. Thanks again for your willingness in debugging and testing the issues at hand!

**Comment by Brooke-white** [comment]
These fixes were included in release v2.0.885 made today. Please re-open this issue if you continue to experience this issue after using the latest release.
